src/Common/SettingsChanges.h

```diff
@@ -1,24 +1,50 @@
 #pragma once
 
 #include <Core/Field.h>
+#include <Parsers/IAST.h>
 
 
 namespace DB
 {
 
 class IColumn;
 
-struct SettingChange
+class SettingChange
 {
+private:
     String name;
-    Field value;
+    Field field_value;
+    ASTPtr ast_value; /// A setting value which cannot be put in Field.
```

Comment:
On second thought maybe it wasn't very good idea to add `ASTPtr` here because `SettingChange` lives in `src/Common` and `ASTPtr` is parser's stuff, so it looks like we're going to make our dependencies worse. Probably it would be better if we could either extend `Field` to support function-like syntax `name(field, field, field)`, or invent some wrapper like [ASTDictionarySettings](https://github.com/ClickHouse/ClickHouse/blob/c003c9034174b0c0396f0a4ca7fff5706b6cbcf2/src/Parsers/ASTDictionary.h#L69).

Reply:
> How will it help if this wrapper will also be in Parsers?`ASTStorage` contains the field `settings`:

```
ASTSetQuery * settings
```

and we could change `ASTSetQuery` to support extra fields of `ASTPtr` type; and then we would have to write some extra code in order to correctly convert that new `ASTSetQuery` to `MergeTreeSettings`.

Reply:
I don't insist on this way, if everything is ready we can merge it now and (maybe) do some improvements later.